### PR TITLE
comment out definition of ARMA_NO_DEBUG

### PR DESCRIPTION
**Pull request by conradsnicta**
@rcurtin related to discussion in #3300

### CODE REVIEW

**Comment by rcurtin** [comment]
Thanks---I wanted to do some benchmarking before pulling the trigger on this.  For machine learning algorithms that are linear-algebra heavy, I'm not too worried, but we also have lots of algorithms that do a lot of direct accesses to Armadillo matrices.  So, I wanted to see if there was any effect on those.  It turns out that there actually is, so I need to do a little auditing of the tree-building code:

On master:

```
$ for i in `seq 1 5`; do bin/mlpack_fastmks -r ~/datasets/covertype.csv -k 5 -v | grep tree_building; done
[INFO ]   tree_building: 40.978713s
[INFO ]   tree_building: 43.269902s
[INFO ]   tree_building: 40.361299s
[INFO ]   tree_building: 41.638788s
[INFO ]   tree_building: 41.748335s
```

On this branch:

```
$ for i in `seq 1 5`; do bin/mlpack_fastmks -r ~/datasets/covertype.csv -k 5 -v | grep tree_building; done
[INFO ]   tree_building: 50.693736s
[INFO ]   tree_building: 49.618033s
[INFO ]   tree_building: 49.679365s
[INFO ]   tree_building: 49.718268s
[INFO ]   tree_building: 50.417402s
```

I checked a number of other things and they looked good, but I need to do a more exhaustive survey.  First I'll look into the issue above.

**Comment by conradsnicta** [comment]
@rcurtin I'm ok to close/cancel this PR for now, and postpone it until mlpack 4.1 or later.  

mlpack appears to work okay without this PR. Resolving the speed issues with the debugging enabled may require too much effort/time, potentially delaying mlpack 4.0.1 and/or the associated JOSS article.  This PR is more of a "nice to have" aspect rather than something super critical.

**Comment by rcurtin** [comment]
No need to close this---let me work through a couple common cases and see what I find.  I don't think this change holds up either the 4.0.1 release or the JOSS paper.  I also don't mind if there is some slight slowdown, but I want to avoid anything significant.  We can always additionally add a note to the README indicating that once a program is fully debugged, one could set `ARMA_NO_DEBUG` to potentially get some extra speedup (at the cost of slightly less safety).

File: CMakeLists.txt [hunk 1]
```diff
@@ -246,7 +246,7 @@ if (DEBUG)
     endif()
   endif()
 else()
-  add_definitions(-DARMA_NO_DEBUG)
+  # add_definitions(-DARMA_NO_DEBUG)
```
Comment (rcurtin) [inline CMakeLists.txt:249]:
```suggestion
```

**Review by rcurtin — APPROVED** [review]
Okay, it took a long time but I did a much more thorough benchmarking.  I noticed when I was chasing down perceived slowdowns that they were mysterious and not reproducible.  I think I was observing benchmarking noise.  So, I decided to find a dedicated system I was not also actively using to, I dunno, browse the internet or whatever, and ran a whole battery of simulations.  10 trials each, I compared to see if there was any serious difference.  In general there is not, or at least, there is not enough of a difference that I think is significant.  Due to what I take to be random benchmarking noise, some of the runs were actually observed to be faster with this patch, but again, my take is that the speedup is not actually the result of the code change but instead random noise.

There is *possibly* one exception that I observed in the `FastMKS` code, which builds cover trees (a complicated process).  This code frequently calls `dataset->col(point)`:

https://github.com/mlpack/mlpack/blob/29479d78140a7994909fe35c47518ab178f3a8e8/src/mlpack/core/tree/cover_tree/cover_tree_impl.hpp#L1396-L1400

`Mat<eT>::col()` does have a bounds check in it that would be disabled by `ARMA_NO_DEBUG`.  Of course such a check is not computationally intensive, but I think this may be the source of the only reproducible slowdown I found in my benchmarking.  That slowdown isn't enough to make me want to avoid this PR, but I would be interested to see if there are simple workarounds (I am not sure there are).  That's very low priority though.

For the sake of completeness (because this comment isn't long enough already), here are the complete results of my benchmarking.  I did 10 trials of each method, sometimes with different options, and took the mean of the results.  Sometimes I did one dataset for a method, sometimes more, if I thought it was worth further investigation.  We can call this benchmarking process "mildly principled".

```
mlpack_knn:

  covertype
   * ARMA_NO_DEBUG:   7.4526450s
   * conrad's patch:  7.2516502s
   
  mnist
   * ARMA_NO_DEBUG:   176.6952357s
   * conrad's patch:  175.8793206s
  
  lcdm
   * ARMA_NO_DEBUG:   439.3120701s
   * conrad's patch:  440.6140751s
  
mlpack_adaboost
   
  mnist
   * ARMA_NO_DEBUG:   12.26016850s
   * conrad's patch:  12.25567150s

  pokerhand
   * ARMA_NO_DEBUG:   5.07380689s
   * conrad's patch:  5.22549300s

mlpack_cf

  movielens-100k
   * ARMA_NO_DEBUG:   10.37032830s
   * conrad's patcH:  10.35786060s

mlpack_range_search

  covertype
   * ARMA_NO_DEBUG:   42.9251321s
   * conrad's patch:  42.6008236s

mlpack_kde

  higgs-200k
   * ARMA_NO_DEBUG:   1661.8267836s
   * conrad's patch:  1667.1254135s

mlpack_logistic_regression

  pokerhand 0/1:
   * ARMA_NO_DEBUG:   3.0225906s
   * conrad's patch:  3.0605039s

  pokerhand 0/1 sgd:
   * ARMA_NO_DEBUG:   1.8886423s
   * conrad's patch:  1.9190333s

  covertype 0/1
   * ARMA_NO_DEBUG:   592.63996070s  (outlier on the high end: 621s, without, mean is 589.467606)
   * conrad's patch:  589.2335323s

mlpack_random_forest

  covertype
   * ARMA_NO_DEBUG:   10.0463702s
   * conrad's patch:  10.1587119s

  mnist
   * ARMA_NO_DEBUG:   41.7480111s
   * conrad's patch:  41.9480747s

mlpack_sparse_coding

  mnist
   * ARMA_NO_DEBUG:   1285.2681734s
   * conrad's patch:  1283.2479317s

mlpack_det

  higgs-200k
   * ARMA_NO_DEBUG:   319.49070889s  (also lots of variance, high 322, low 299, next low 318)
   * conrad's patch:  311.05549677s  (lots of variance, high 319, low 273, next low 310)

mlpack_krann

  covertype
   * ARMA_NO_DEBUG:   192.5587985s
   * conrad's patch:  192.7225008s

mlpack_fastmks

  covertype, linear kernel
   * ARMA_NO_DEBUG:   83.7620034s
   * conrad's patch:  84.2836036s

  favorita-10k, gaussian kernel
   * ARMA_NO_DEBUG:   10.9224405s
   * conrad's patch:  11.6948914s
```

Also, we should update HISTORY.md for this change.  I can do that during merge or you can push a commit, up to you.

**Pull request merged**